package.json

```diff
@@ -27,22 +27,40 @@
   "contributes": {
     "commands": [
       {
-        "command": "styledComponentsExtractor.extract",
-        "title": "Extract styled-components"
+        "command": "styledComponentsExtractor.extractToClipboard",
+        "title": "Extract styled-components to clipboard"
       },
       {
-        "command": "styledComponentsExtractor.extractExported",
-        "title": "Extract exported styled-components"
+        "command": "styledComponentsExtractor.extractExportedToClipboard",
+        "title": "Extract exported styled-components to clipboard"
+      },
+      {
+        "command": "styledComponentsExtractor.extractToSameFile",
+        "title": "Extract styled-components to the same file"
+      },
+      {
+        "command": "styledComponentsExtractor.extractToSeparateFile",
+        "title": "Extract exported styled-components to a separate file"
       }
     ],
     "configuration": {
       "type": "object",
       "title": "Styled-Components Extractor configuration",
       "properties": {
+        "styledComponentsExtractor.separateFile.matchPattern": {
+          "type": "string",
+          "default": "*",
+          "markdownDescription": "Matches files that are eligible for extraction. The match pattern excludes the extension. You can use `*` as a wildcard. Only applies when extracting to a separate file."
+        },
+        "styledComponentsExtractor.separateFile.outputFile": {
+          "type": "string",
+          "default": "styles.ts",
+          "markdownDescription": "The name of the file, excluding the extension, to place extracted components extracting to a separate file. You can use `$1`, `$2`, etc to match the contents of each wildcard from `matchPattern`, allowing you to name the file based on the file the component was extracted from."
```

Comment:
If I understand correctly (after reading code in `corresponding-file.ts`), these two configs are supposed to be used together like this:


```js 
{
  matchPattern: '*.*',
  outputFile: '$1.styles.ts'
}
```

It's a powerful API that enables many user cases than alternatives, but it's too clever/implicit IMHO.
It mangles these two things together and did it in an implicit way (that I have to look at code to understand how they pieces together):

1. Act like a white-list for extraction to work
2. (Can be used to) interpolate output file names from part of input file names.

The first thing it does looks unneccessary to me, as it's preventing user extracting when user **actively** triggers the command.
The secodnd

Reply:
I agree with you to some degree! And I was struggling with this a bit myself. I think the biggest issue is that 80+ percent of people should never touch `matchPattern` because most use cases just involve appending a suffix to the filename. 

Here's my suggestion, inspired by yours:
1. We should make it clearer that `matchPattern` is for advanced use cases, potentially renaming it as you suggest. We can also put it in an "advanced" category in the settings. Another low hanging fruit is to explicitly order the settings so `matchPattern` is last.

2. For simple use cases, we should expose a `$filename` variable to the `outputFile` option. This will always match the entire filename except the extension (should we call it `$basename` ?). Most use cases can just use this variable, e.g. `$filename.styles.ts`

3. The default value of `matchPattern`/`inputFilePattern` can be changed to blank/null, making it clear you don't need to use this in most cases. This doesn't affect `$filename`, so this works well for most use cases. If you _do_ specify something here, it works as before, and wildcard-matching values are exposed as `$1`, `$2`, etc.

4. (maybe) We can change `matchPattern`/`inputFilePattern` to be garden-variety regex syntax as you suggest, which is more powerful but slightly more confusing than the glob syntax I proposed. If you really like the regex idea, I have no problem with this!

5. (maybe) If you feel strongly `%1` is better than `$1` we can change all variables to use this.

 I think the only thing that you suggested that I generally (respectfully) disagree with is making inclusion of the extension explicit (i.e. `%ext`) in your proposal. IMHO it would always be an error for the user not to include this variable in their pattern, so I think it's better to always implicitly add the extension rather than allowing the user to make a mistake by omitting it. Of course, as this behavior is implicit., we must document this very carefully.

On top of all this, I was planning to include a "cookbook" of examples in the README to help users understand how to leverage this feature.

Let me know your thoughts! I don't think it would much effort for me to make these changes if they make sense.

Reply:
I agree with you on most points! It would be good if config can be easy to understand/write for simple cases (prefix apppending), and can be smoothly extended to support advanced cases.

So here is an updated (suggested) config inspired by your suggestion:

```js
       "styledComponentsExtractor.separateFile.inputFileRegex": { 
          "type": "string",
          "default": "^(.+)\\.(js|ts|jsx|tsx)$", 
          "markdownDescription": "The regex pattern used to match on input file name. If provided, the match result ($1, $2) can be used in output file name."

          // Considerations:
          // 1. A sensible default is provided here both as teaching material and a working draft that user can trust and start edit. Note that `\\` double escaping is both tricky and important to know.
          // 2. If regex doesn't match, I prefer setting $1, $2, ... to empty string in output, instead of showing error and prevent user from extracting. It's better to be tolerant for this scenario.
          // 3. Invalid regex can throw error, we should catch it (and show error?)
          // 4. Regex vs glob:  regex is preferred over glob syntax here, as regex and $1, $2... are integrated concepts. Using glob would require some implicit conversion thus harder to understand.  Also regex is more powerful for advanced cases that we are fulfilling here.
      },


      "styledComponentsExtractor.separateFile.outputFile": {
          "type": "string",
          "default": "$name.styles.$ext",
          "markdownDescription": "The name of the file to place extracted components. $name and $ext will be replaced by the base name and extension of input file respectively. If `styledComponentsExtractor.separateFile.inputFileRegex` is provided, the matching result ($1, $2...) can also be used."

         // Considerations:
         // 1. Easy to understand from first look, straightly mapping to the simple cases users are expecting
         // 2. Can be smoothly upgraded to support advanced cases (by adding $1, $2 using `inputFileRegex`)
      },

```

As for each of your points:

1. Agree that `matchPattern` (or inputFileRegex) should be optional and act as an advanced config, but I won't bother adjusting order of config, default alphabetical order is okay to me :)
2. Agree, and $name/$basename/$base are all acceptable, I used $name because [path](https://devdocs.io/node/path#pathparsepath)
3. Agree, but a sensible default can be provided as reasons stated above.
4. I still prefer regex here, as it's conceptional integrated with $1, $2... . Glob syntax needs some "magical" conversion to work.
5. I originally suggested %1, %2 to avoid confusion with $1, $2 to be added later, but inspired by your suggestion, I think $base/$ext is better, and for advanced cases we should just go with regex as your first API suggests.
6. As for implicit $ext and explicit $ext, I personally like explicit one, as it's: 
    1. conceptionally simpler. `$name.styles.$ext` is quite straight forward to me), this can avoids user guessing.
    2. along with $1, $2, user could use its own advanced logic to determine on $ext, so we don't stand in the way by force adding an ext.
    3. In future, we might add even features like `auto extract unbound tags to React Component`, who knows ;-)
    That's said, if you strongly prefer auto-adding $ext, I'm okay with that! As long as this behavior is documented.

What do you think about suggested config , also please feel free to modify any wording/variable names as I'm not fluent in English :)

Reply:
Just pushed an update which should be very close to making both of us happy, with luck!
- Configuration is called `inputFileRegex` now.
- It is a full blown regex and not a glob.
- Default for that is blank, to emphasize its optionality. You had suggested an actual regex as the default, as a hopefully acceptable compromise, that regex is relocated to the documentation in the README file (I agree that an example with escaping is essential).
- `$name` variable now available regardless of whether you set `inputFileRegex`. Agreed with using `path` as the source of truth, I had intended the same, but got confused originally.
- Still not using explicit `$ext`, I have tried to document this as much as I can.
- README is updated, including a set of examples on the match patterns.

Reply:
Fantastic! Really appreciate meaningful discussion and final result. Let's go with this.